### PR TITLE
Use java.time in Headers

### PR DESCRIPTION
**Pull request by swankjesse**
Closes: https://github.com/square/okhttp/issues/3921

### CODE REVIEW

**Comment by swankjesse** [comment]
Build flake:

```

[ERROR] Tests run: 90, Failures: 3, Errors: 0, Skipped: 3, Time elapsed: 28.003 s <<< FAILURE! - in okhttp3.internal.http2.HttpOverHttp2Test
[ERROR] cancelWithStreamNotCompleted[h2_prior_knowledge](okhttp3.internal.http2.HttpOverHttp2Test)  Time elapsed: 0.005 s  <<< FAILURE!
java.lang.AssertionError: expected:<0> but was:<1>
	at okhttp3.internal.http2.HttpOverHttp2Test.cancelWithStreamNotCompleted(HttpOverHttp2Test.java:758)

[ERROR] concurrentHttp2ConnectionsDeduplicated[h2](okhttp3.internal.http2.HttpOverHttp2Test)  Time elapsed: 0.21 s  <<< FAILURE!
java.lang.AssertionError: expected:<1> but was:<2>
	at okhttp3.internal.http2.HttpOverHttp2Test.concurrentHttp2ConnectionsDeduplicated(HttpOverHttp2Test.java:1420)

[ERROR] cancelWithStreamNotCompleted[h2](okhttp3.internal.http2.HttpOverHttp2Test)  Time elapsed: 0.001 s  <<< FAILURE!
java.lang.AssertionError: expected:<0> but was:<1>
	at okhttp3.internal.http2.HttpOverHttp2Test.cancelWithStreamNotCompleted(HttpOverHttp2Test.java:758)
```

File: okhttp/src/main/java/okhttp3/Headers.java [hunk 1]
```diff
@@ -334,15 +346,25 @@ public Builder addAll(Headers headers) {
...
       add(name, HttpDate.format(value));
       return this;
     }
 
+    /**
+     * Add a header with the specified name and formatted instant. Does validation of header names
+     * and value.
+     */
+    @IgnoreJRERequirement
+    public Builder add(String name, Instant value) {
```
[the remark this answers is not in the document]
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/Headers.java]: Good insight, and spot on.

```
[INFO] --- maven-compiler-plugin:3.8.0:compile (default-compile) @ overload7 ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 1 source file to /Users/jwilson/Projects/overload7/target/classes
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR :
[INFO] -------------------------------------------------------------
[ERROR] /Users/jwilson/Projects/overload7/src/main/java/com/publicobject/overload7/App.java:[9,9] cannot access java.time.Instant
  class file for java.time.Instant not found
[INFO] 1 error
```

Well shit.

**Comment by swankjesse** [comment]
My options as I see ’em:

**A.** Merge this now. For Java it won’t be a problem – we’re requiring Java 8 now! For Android I expect _most_ developers to use a target SDK that includes `java.time`.

**B.** Merge this later. API 26 came out in August 2017 and might not be widely used as a target SDK.

**C.** Rename the methods. `setInstant()` and `addInstant()` work. It’s the strategy that’s most robust for the future.

Advice?

**Comment by JakeWharton** [comment]
It's the compile SDK version that's needed, and I think it's reasonable to say OkHttp version X requires compilation with Android API version 26 or newer. It's not dissimilar from saying OkHttp version X requires Java 8. Plus I'm in favor of dragging the Android ecosystem forward and the use of Java 8 bytecode and APIs means up-to-date compile SDKs and toolchains which benefit every library.

**Comment by digitalbuddha** [comment]
"Google Play will require that new apps target at least Android 8.0 (API level 26) from August 1, 2018, and that app updates target Android 8.0 from November 1, 2018." you should be safe

**Comment by JakeWharton** [comment]
Nope I'm on board!

**Comment by swankjesse** [comment]
✔️ ?

**Pull request merged**